File: backend/sources/open_meteo.py
import httpx
from .base import WeatherSource, wmo_to_condition, get_condition_icon
from config import SOURCE_TIMEOUT, TIMEZONE


class OpenMeteoSource(WeatherSource):
    name = "open_meteo"
    BASE_URL = "https://api.open-meteo.com/v1/forecast"

    async def fetch_current(self, lat: float, lon: float) -> dict:
        params = {
            "latitude": lat,
            "longitude": lon,
            "current": "temperature_2m,relative_humidity_2m,wind_speed_10m,wind_direction_10m,precipitation,weather_code",
            "timezone": TIMEZONE,
        }
        async with httpx.AsyncClient(timeout=SOURCE_TIMEOUT) as client:
            resp = await client.get(self.BASE_URL, params=params)
            resp.raise_for_status()
            data = resp.json()

        current = data["current"]
        condition = wmo_to_condition(current.get("weather_code", 0))
        return {
            "temperature": current.get("temperature_2m", 0),
            "humidity": int(current.get("relative_humidity_2m", 0)),
            "wind_speed": current.get("wind_speed_10m", 0),
            "wind_direction": int(current.get("wind_direction_10m", 0)),
            "precipitation": current.get("precipitation", 0),
            "condition": condition,
            "condition_icon": get_condition_icon(condition),
        }

    async def fetch_hourly(self, lat: float, lon: float) -> list[dict]:
        params = {
            "latitude": lat,
            "longitude": lon,
            "hourly": "temperature_2m,relative_humidity_2m,wind_speed_10m,precipitation,weather_code",
            "timezone": TIMEZONE,
            "forecast_hours": 24,
        }
        async with httpx.AsyncClient(timeout=SOURCE_TIMEOUT) as client:
            resp = await client.get(self.BASE_URL, params=params)
            resp.raise_for_status()
            data = resp.json()

        hourly = data.get("hourly", {})
        times = hourly.get("time", [])
        results = []
        for i, time_str in enumerate(times):
            condition = wmo_to_condition(hourly["weather_code"][i]) if hourly.get("weather_code") else "necunoscut"
            hour_part = time_str.split("T")[1][:5] if "T" in time_str else time_str
            results.append({
                "hour": hour_part,
                "timestamp": time_str,
                "temperature": hourly["temperature_2m"][i],
                "humidity": int(hourly["relative_humidity_2m"][i]),
                "wind_speed": hourly["wind_speed_10m"][i],
                "precipitation": hourly.get("precipitation", [0] * len(times))[i],
                "condition": condition,
            })
        return results

    async def fetch_daily(self, lat: float, lon: float) -> list[dict]:
        params = {
            "latitude": lat,
            "longitude": lon,
            "daily": "temperature_2m_max,temperature_2m_min,weather_code,precipitation_sum,wind_speed_10m_max",
            "timezone": TIMEZONE,
            "forecast_days": 7,
        }
        async with httpx.AsyncClient(timeout=SOURCE_TIMEOUT) as client:
            resp = await client.get(self.BASE_URL, params=params)
            resp.raise_for_status()
            data = resp.json()

        daily = data.get("daily", {})
        dates = daily.get("time", [])
        day_names_ro = ["Luni", "Marți", "Miercuri", "Joi", "Vineri", "Sâmbătă", "Duminică"]
        results = []
        for i, date_str in enumerate(dates):
            from datetime import datetime
            dt = datetime.strptime(date_str, "%Y-%m-%d")
            day_name = day_names_ro[dt.weekday()]
            condition = wmo_to_condition(daily["weather_code"][i]) if daily.get("weather_code") else "necunoscut"
            results.append({
                "date": date_str,
                "day_name": day_name,
                "temp_min": daily["temperature_2m_min"][i],
                "temp_max": daily["temperature_2m_max"][i],
                "precipitation": daily.get("precipitation_sum", [0] * len(dates))[i],
                "wind_speed": daily.get("wind_speed_10m_max", [0] * len(dates))[i],
                "condition": condition,
            })
        return results

File: src/lib/sources/open-meteo.ts
import {
  WeatherSource,
  CurrentData,
  HourlyData,
  DailyData,
  wmoToCondition,
  getConditionIcon,
  getDayNameRo,
} from "./base";

const BASE_URL = "https://api.open-meteo.com/v1/forecast";
const TIMEOUT = 15000;

export class OpenMeteoSource implements WeatherSource {
  name = "open_meteo";

  async fetchCurrent(lat: number, lon: number): Promise<CurrentData> {
    const params = new URLSearchParams({
      latitude: lat.toString(),
      longitude: lon.toString(),
      current:
        "temperature_2m,relative_humidity_2m,wind_speed_10m,wind_direction_10m,precipitation,weather_code",
      timezone: "Europe/Bucharest",
    });

    const resp = await fetch(`${BASE_URL}?${params}`, {
      signal: AbortSignal.timeout(TIMEOUT),
    });
    const data = await resp.json();
    const current = data.current;
    const condition = wmoToCondition(current?.weather_code ?? 0);

    return {
      temperature: current?.temperature_2m ?? 0,
      humidity: Math.round(current?.relative_humidity_2m ?? 0),
      wind_speed: current?.wind_speed_10m ?? 0,
      wind_direction: Math.round(current?.wind_direction_10m ?? 0),
      precipitation: current?.precipitation ?? 0,
      condition,
      condition_icon: getConditionIcon(condition),
    };
  }

  async fetchHourly(lat: number, lon: number): Promise<HourlyData[]> {
    const params = new URLSearchParams({
      latitude: lat.toString(),
      longitude: lon.toString(),
      hourly:
        "temperature_2m,relative_humidity_2m,wind_speed_10m,precipitation,weather_code",
      timezone: "Europe/Bucharest",
      forecast_hours: "24",
    });

    const resp = await fetch(`${BASE_URL}?${params}`, {
      signal: AbortSignal.timeout(TIMEOUT),
    });
    const data = await resp.json();
    const hourly = data.hourly || {};
    const times: string[] = hourly.time || [];

    return times.map((t, i) => {
      const condition = wmoToCondition(hourly.weather_code?.[i] ?? 0);
      return {
        hour: t.includes("T") ? t.split("T")[1].slice(0, 5) : t,
        timestamp: t,
        temperature: hourly.temperature_2m?.[i] ?? 0,
        humidity: Math.round(hourly.relative_humidity_2m?.[i] ?? 0),
        wind_speed: hourly.wind_speed_10m?.[i] ?? 0,
        precipitation: hourly.precipitation?.[i] ?? 0,
        condition,
      };
    });
  }

  async fetchDaily(lat: number, lon: number): Promise<DailyData[]> {
    const params = new URLSearchParams({
      latitude: lat.toString(),
      longitude: lon.toString(),
      daily:
        "temperature_2m_max,temperature_2m_min,weather_code,precipitation_sum,wind_speed_10m_max",
      hourly: "relative_humidity_2m",
      timezone: "Europe/Bucharest",
      forecast_days: "7",
    });

    const resp = await fetch(`${BASE_URL}?${params}`, {
      signal: AbortSignal.timeout(TIMEOUT),
    });
    const data = await resp.json();
    const daily = data.daily || {};
    const hourly = data.hourly || {};
    const dates: string[] = daily.time || [];

    // Calculeaza media humidity pe zi din date orare
    const hourlyTimes: string[] = hourly.time || [];
    const hourlyHumidity: number[] = hourly.relative_humidity_2m || [];
    const humidityByDay: Record<string, number[]> = {};
    for (let i = 0; i < hourlyTimes.length; i++) {
      const dateKey = hourlyTimes[i].split("T")[0];
      if (!humidityByDay[dateKey]) humidityByDay[dateKey] = [];
      if (typeof hourlyHumidity[i] === "number") {
        humidityByDay[dateKey].push(hourlyHumidity[i]);
      }
    }

    return dates.map((d, i) => {
      const dayHumidity = humidityByDay[d] || [];
      const avgHumidity =
        dayHumidity.length > 0
          ? Math.round(
              dayHumidity.reduce((a, b) => a + b, 0) / dayHumidity.length,
            )
          : undefined;

      return {
        date: d,
        day_name: getDayNameRo(d),
        temp_min: daily.temperature_2m_min?.[i] ?? 0,
        temp_max: daily.temperature_2m_max?.[i] ?? 0,
        humidity: avgHumidity,
        precipitation: daily.precipitation_sum?.[i] ?? 0,
        wind_speed: daily.wind_speed_10m_max?.[i] ?? 0,
        condition: wmoToCondition(daily.weather_code?.[i] ?? 0),
      };
    });
  }
}
